## UN Data Exploration

#### Q1. Download the Gross Domestic Product (GDP) per capita dataset from http://data.un.org/Data.aspx?d=WDI&f=Indicator_Code%3aNY.GDP.PCAP.PP.KD. Rename it to gdp_per_capita.csv and place it in the data folder of your project repository.

#### Q2. Create a Jupyter Notebook in the notebooks folder and name it UN_Data_Exploration.

#### Q3. In the first cell of your notebook, import the required packages with their customary aliases as follows:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#### Q4. Using the pandas read_csv() function, read the GDP dataset into your notebook as a DataFrame called gdp_df. After reading it in, inspect the first 10 rows and then inspect the last 10 rows

In [3]:
gdp_df = pd.read_csv('../data/gdp_per_capita.csv')
gdp_df

Country or Area  Year        Value  Value Footnotes
0        Afghanistan  2021  1517.016266              NaN
1        Afghanistan  2020  1968.341002              NaN
2        Afghanistan  2019  2079.921861              NaN
3        Afghanistan  2018  2060.698973              NaN
4        Afghanistan  2017  2096.093111              NaN
...              ...   ...          ...              ...
7657        Zimbabwe  1994  2670.106615              NaN
7658        Zimbabwe  1993  2458.783255              NaN
7659        Zimbabwe  1992  2468.278257              NaN
7660        Zimbabwe  1991  2781.787843              NaN
7661        Zimbabwe  1990  2704.757299              NaN

[7662 rows x 4 columns]

In [ ]:
gdp_df.head(10)

In [ ]:
gdp_df.tail(10)

#### Q5. Drop the 'Value Footnotes' column, and rename the remaining columns to 'Country', 'Year', and 'GDP_Per_Capita'.

In [5]:
#drop column: df.drop(columns=['col_name'])
#rename column: df.rename(columns={""oldname": "newname"})
gdp_df = (
    gdp_df
    .drop(columns=['Value Footnotes'])
    .rename(columns={"Country or Area": "Country", "Value": "GDP_Per_Capita"})
)
gdp_df

Country  Year  GDP_Per_Capita
0     Afghanistan  2021     1517.016266
1     Afghanistan  2020     1968.341002
2     Afghanistan  2019     2079.921861
3     Afghanistan  2018     2060.698973
4     Afghanistan  2017     2096.093111
...           ...   ...             ...
7657     Zimbabwe  1994     2670.106615
7658     Zimbabwe  1993     2458.783255
7659     Zimbabwe  1992     2468.278257
7660     Zimbabwe  1991     2781.787843
7661     Zimbabwe  1990     2704.757299

[7662 rows x 3 columns]

#### Q6. How many rows and columns does gdp_df have? What are the data types of its columns? If any of the columns are not the expected types, figure out why and fix it.

7662 rows and 3 columns:
1. Country - object
2. Year - integer
3. GDP_Per_Capita - float

In [ ]:
gdp_df.shape

In [ ]:
gdp_df.info()
#gdp_df.dtypes

#### Q7. Which years are represented in this dataset? Take a look at the number of observations per year. What do you notice?

1990 - 2022 (2022 is hidden between 2001 and 2002!)  
The number of observations per year increases over time.

In [ ]:
gdp_df['Year'].unique()

In [11]:
#gdp_df['Year'].value_counts(sort=False)
gdp_df['Year'].value_counts().sort_index(ascending=False)

Year
2022    232
2021    241
2020    242
2019    242
2018    242
2017    242
2016    242
2015    242
2014    242
2013    242
2012    240
2011    240
2010    239
2009    239
2008    238
2007    237
2006    237
2005    236
2004    236
2003    235
2002    235
2001    234
2000    233
1999    227
1998    226
1997    226
1996    223
1995    223
1994    213
1993    211
1992    210
1991    208
1990    207
Name: count, dtype: int64

#### Q8. How many countries are represented in this dataset? Which countries are least represented in the dataset? Why do you think these countries have so few observations?

242 Countries (and Regions)  
Djibouti and Somalia have the fewest entries; they didn't start reporting or tracking until 2013 (maybe it's voluntary?)?

In [ ]:
gdp_df['Country'].nunique()

In [ ]:
gdp_df['Country'].unique()

In [13]:
gdp_df['Country'].value_counts()

Country
Least developed countries: UN classification    33
Middle East & North Africa                      33
Middle East & North Africa (IDA & IBRD)         33
Middle income                                   33
Mongolia                                        33
                                                ..
Kosovo                                          15
Sint Maarten (Dutch part)                       14
Turks and Caicos Islands                        12
Somalia                                         10
Djibouti                                        10
Name: count, Length: 242, dtype: int64

In [17]:
gdp_df['Country'].value_counts().value_counts(sort=False)

count
33    202
32      5
31      3
30      1
29      2
28     10
26      2
25      1
24      1
23      4
22      3
20      1
19      1
16      1
15      1
14      1
12      1
10      2
Name: count, dtype: int64

In [ ]:
gdp_df.loc[gdp_df['Country'] == 'Djibouti']

In [ ]:
gdp_df.loc[gdp_df['Country'] == 'Somalia']

In [ ]:
gdp_df.loc[gdp_df['Country'] == 'Afghanistan']

#### Q9. Create a new dataframe by subsetting gdp_df to just the year 2021. Call this new dataframe gdp_2021

In [21]:
gdp_2021 = gdp_df.loc[gdp_df['Year'] == 2021]
#gdp_2021 = gdp_df[gdp_df['Year'] == 2021]
gdp_2021

Country  Year  GDP_Per_Capita
0                     Afghanistan  2021     1517.016266
21    Africa Eastern and Southern  2021     3519.174840
54     Africa Western and Central  2021     4014.607965
87                        Albania  2021    14595.944386
120                       Algeria  2021    11029.138782
...                           ...   ...             ...
7502                     Viet Nam  2021    10628.219166
7535           West Bank and Gaza  2021     5641.044400
7564                        World  2021    17055.357429
7597                       Zambia  2021     3236.788981
7630                     Zimbabwe  2021     2115.144555

[241 rows x 3 columns]

#### Q10. Use .describe() to find the summary statistics for GDP per capita in 2021.

In [ ]:
gdp_2021['GDP_Per_Capita'].describe()

#### Q11. Create a histogram of GDP Per Capita numbers for 2021 (you may wish to adjust the number of bins for your histogram). How would you describe the shape of the distribution?

The distribution is heavily right-skewed, with most countries/regions at or below a GDP per Capita of 20K.

In [ ]:
fig,ax = plt.subplots(figsize = (10,6))   

plt.hist(
    data=gdp_2021,
    x='GDP_Per_Capita',
    edgecolor='Black',
    linewidth=2,
    bins = 15
    );
plt.xlabel('GDP Per Capita')                            
plt.ylabel('Number of Countries')
plt.title('Distribution of Global GDP Per Capita for 2021');

#### Q12. Find the top 5 countries and bottom 5 countries by GDP per capita in 2021.

In [ ]:
gdp_2021.sort_values('GDP_Per_Capita', ascending = False)

#### Q13. Now, return to the full dataset, gdp_df. Pivot the data for 1990 and 2021 (using the pandas .pivot_table() method or another method) so that each row corresponds to a country, each column corresponds to a year, and the values in the table give the GDP_Per_Capita amount. Drop any rows that are missing values for either 1990 or 2021. Save the result to a dataframe named gdp_pivoted.

In [97]:
gdp_pivotedtotal = gdp_df.pivot_table(values = "GDP_Per_Capita", index = "Country", columns = 'Year')
gdp_pivotedtotal

Year                                1990         1991         1992  \
Country                                                              
Afghanistan                          NaN          NaN          NaN   
Africa Eastern and Southern  3037.297466  2955.642238  2823.940366   
Africa Western and Central   2788.301039  2750.790764  2743.855561   
Albania                      4827.027705  3496.369626  3264.820757   
Algeria                      8828.874473  8517.376962  8471.527605   
...                                  ...          ...          ...   
Viet Nam                     2099.394649  2177.473744  2317.266668   
West Bank and Gaza                   NaN          NaN          NaN   
World                        9705.981267  9669.677060  9665.890260   
Zambia                       2290.039226  2232.837441  2141.504615   
Zimbabwe                     2704.757299  2781.787843  2468.278257   

Year                                1993         1994         1995  \
Country                                                              
Afghanistan                          NaN          NaN          NaN   
Africa Eastern and Southern  2737.731240  2715.131116  2764.305017   
Africa Western and Central   2644.709683  2575.064177  2561.665446   
Albania                      3598.810267  3921.614970  4471.601702   
Algeria                      8109.883559  7869.270272  8013.123442   
...                                  ...          ...          ...   
Viet Nam                     2455.508783  2623.720022  2825.016443   
West Bank and Gaza                   NaN  3951.205493  4047.128488   
World                        9675.232260  9799.764965  9957.172695   
Zambia                       2232.710379  1991.185925  1999.356842   
Zimbabwe                     2458.783255  2670.106615  2641.378271   

Year                                 1996          1997          1998  \
Country                                                                 
Afghanistan                           NaN           NaN           NaN   
Africa Eastern and Southern   2838.692029   2886.566235   2867.960243   
Africa Western and Central    2612.194795   2654.384927   2676.529845   
Albania                       4908.932392   4400.312754   4819.067832   
Algeria                       8195.860480   8147.878198   8435.035658   
...                                   ...           ...           ...   
Viet Nam                      3039.938041   3239.681646   3378.904834   
West Bank and Gaza            3916.925775   4294.746098   4786.480236   
World                        10179.565344  10424.112458  10532.457767   
Zambia                        2071.708828   2096.294593   2034.897183   
Zimbabwe                      2867.026043   2896.147308   2931.725144   

Year                                 1999  ...          2013          2014  \
Country                                    ...                               
Afghanistan                           NaN  ...   2165.340915   2144.449634   
Africa Eastern and Southern   2873.553735  ...   3593.299065   3642.875373   
Africa Western and Central    2649.555854  ...   4026.231916   4146.994622   
Albania                       5474.849914  ...  11361.252492  11586.817446   
Algeria                       8584.071496  ...  11360.637612  11561.259795   
...                                   ...  ...           ...           ...   
Viet Nam                      3495.097968  ...   7257.729273   7641.909252   
West Bank and Gaza            5052.064072  ...   6118.257181   5967.073437   
World                        10754.895302  ...  14801.332173  15120.730322   
Zambia                        2074.453663  ...   3330.876903   3375.941270   
Zimbabwe                      2866.032886  ...   2333.644532   2316.954536   

Year                                 2015          2016          2017  \
Country                                                                 
Afghanistan                   2108.714173   2101.422187   2096.0931

In [101]:
gdp_subset = gdp_df[gdp_df["Year"].isin([1990, 2001])]
gdp_subset

Country  Year  GDP_Per_Capita
41    Africa Eastern and Southern  2001     2928.062946
52    Africa Eastern and Southern  1990     3037.297466
74     Africa Western and Central  2001     2734.257633
85     Africa Western and Central  1990     2788.301039
107                       Albania  2001     6441.440698
...                           ...   ...             ...
7595                        World  1990     9705.981267
7617                       Zambia  2001     2142.787524
7628                       Zambia  1990     2290.039226
7650                     Zimbabwe  2001     2772.325234
7661                     Zimbabwe  1990     2704.757299

[441 rows x 3 columns]

In [103]:
gdp_pivoted = gdp_subset.pivot_table(values = "GDP_Per_Capita", index = "Country", columns = 'Year')
gdp_pivoted

Year                                1990          2001
Country                                               
Africa Eastern and Southern  3037.297466   2928.062946
Africa Western and Central   2788.301039   2734.257633
Albania                      4827.027705   6441.440698
Algeria                      8828.874473   8926.110134
Angola                       5793.084512   4768.008894
...                                  ...           ...
Viet Nam                     2099.394649   3879.338958
West Bank and Gaza                   NaN   3980.933349
World                        9705.981267  11221.662910
Zambia                       2290.039226   2142.787524
Zimbabwe                     2704.757299   2772.325234

[234 rows x 2 columns]

In [105]:
gdp_pivoted = gdp_pivoted.dropna()
gdp_pivoted

Year                                1990          2001
Country                                               
Africa Eastern and Southern  3037.297466   2928.062946
Africa Western and Central   2788.301039   2734.257633
Albania                      4827.027705   6441.440698
Algeria                      8828.874473   8926.110134
Angola                       5793.084512   4768.008894
...                                  ...           ...
Vanuatu                      2774.138350   2782.053642
Viet Nam                     2099.394649   3879.338958
World                        9705.981267  11221.662910
Zambia                       2290.039226   2142.787524
Zimbabwe                     2704.757299   2772.325234

[207 rows x 2 columns]

#### Q14. Create a new column in gdp_pivoted named Percent_Change. This column should contain the percent change in GDP_Per_Capita from 1990 to 2021. Hint: Percent change is calculated as 100*(New Value - Old Value) / Old Value.

In [ ]:
gdp_pivoted['Percent_Change'] = 100*(gdp_pivoted[2001] - gdp_pivoted[1990]) / gdp_pivoted[1990]
gdp_pivoted

#### Q15. How many countries experienced a negative percent change in GDP per capita from 1990 to 2021?

In [ ]:
gdp_pivoted['Percent_Change'].agg(lambda x: sum(x < 0))

#### Q16. Which country had the highest % change in GDP per capita? Create a line plot showing this country's GDP per capita for all years from 1990 to 2018. Create another showing the country with the second highest % change in GDP. How do the trends in these countries compare?

Equatorial Guinea	1543.403611%  
China	160.716880%  
EG GDP per capita has a meteoric rise until about 2009, at which point it begins to trends downward; whereas China trends upward more gradually

In [ ]:
gdp_pivoted.sort_values('Percent_Change', ascending = False)

#### Q16 Bonus: Put both line charts on the same plot.

In [ ]:
gdp_top2 = gdp_df[gdp_df['Country'].isin(['Equatorial Guinea', 'China'])]
gdp_top2 = gdp_top2[gdp_top2['Year'] < 2019]
gdp_top2

In [ ]:
gdp_pivoted_top2 = gdp_top2.pivot_table(values = "GDP_Per_Capita", index = 'Year', columns = 'Country')
gdp_pivoted_top2

In [ ]:
gdp_pivoted_top2.plot()
plt.title('GDP over Time')
plt.ylabel('GDP');

In [ ]:
gdp_df_top2.groupby('Country').plot(x='Year')

#### Q17. Read in continents.csv contained in the data folder into a new dataframe called continents. We will be using this dataframe to add a new column to our dataset.

In [107]:
continents = pd.read_csv('../data/continents.csv')
continents

Continent             Country
0        Asia         Afghanistan
1      Europe             Albania
2      Africa             Algeria
3      Europe             Andorra
4      Africa              Angola
..        ...                 ...
211      Asia             Vietnam
212      Asia  West Bank and Gaza
213      Asia               Yemen
214    Africa              Zambia
215    Africa            Zimbabwe

[216 rows x 2 columns]

#### Q18. Merge gdp_df and continents. Keep only the countries that appear in both data frames. Save the result back to gdp_df.

In [109]:
gdp_df = pd.merge(gdp_df, continents, on = 'Country', how = 'inner', validate='many_to_one')
gdp_df

Country  Year  GDP_Per_Capita Continent
0     Afghanistan  2021     1517.016266      Asia
1     Afghanistan  2020     1968.341002      Asia
2     Afghanistan  2019     2079.921861      Asia
3     Afghanistan  2018     2060.698973      Asia
4     Afghanistan  2017     2096.093111      Asia
...           ...   ...             ...       ...
5888     Zimbabwe  1994     2670.106615    Africa
5889     Zimbabwe  1993     2458.783255    Africa
5890     Zimbabwe  1992     2468.278257    Africa
5891     Zimbabwe  1991     2781.787843    Africa
5892     Zimbabwe  1990     2704.757299    Africa

[5893 rows x 4 columns]

#### Q19. Determine the number of countries per continent. Create a bar chart showing this.

In [ ]:
gdp_df['Country'].nunique()

In [ ]:
co_subset = gdp_df[['Country', 'Continent']]
co_subset

In [ ]:
co_subset = co_subset.drop_duplicates(subset=['Country'])
co_subset

In [ ]:
co_subset['Continent'].value_counts().plot(kind='bar')
plt.xticks(rotation=0)
plt.title('Number of Countries by Continent');

#### Q20. Create a seaborn boxplot showing GDP per capita in 2021 split out by continent. What do you notice?

Europe and Asia have a lot more variability, and a couple of pretty extreme outliers; Africa, despite having more countries that any other continent has very little variability, and is also at the lower end for GDP per capita

In [ ]:
gdp_subset2021 = gdp_df.loc[gdp_df['Year'] == 2021]
gdp_subset2021

In [ ]:
plt.figure(figsize = (10,8))
sns.boxplot(data = gdp_subset2021.sort_values('Continent'), x = 'GDP_Per_Capita', y ='Continent');

#### Q21. Download the full csv containing Life expectancy at birth, total (years) from https://data.worldbank.org/indicator/SP.DYN.LE00.IN?name_desc=false. Read this data into a DataFrame named life_expectancy. Note: When reading this dataset it, you may encounter an error. Modify your read_csv call to correct this without modifying the original csv file.


In [151]:
#skiprows argument added to skip over 'header' rows that were messing up the parsing of the dataset
life_expectancy = pd.read_csv('../data/API_SP.DYN.LE00.IN_DS2_en_csv_v2_31632/API_SP.DYN.LE00.IN_DS2_en_csv_v2_31632.csv', skiprows=4)
life_expectancy

Country Name Country Code  \
0                          Aruba          ABW   
1    Africa Eastern and Southern          AFE   
2                    Afghanistan          AFG   
3     Africa Western and Central          AFW   
4                         Angola          AGO   
..                           ...          ...   
261                       Kosovo          XKX   
262                  Yemen, Rep.          YEM   
263                 South Africa          ZAF   
264                       Zambia          ZMB   
265                     Zimbabwe          ZWE   

                              Indicator Name  Indicator Code       1960  \
0    Life expectancy at birth, total (years)  SP.DYN.LE00.IN  64.152000   
1    Life expectancy at birth, total (years)  SP.DYN.LE00.IN  44.085552   
2    Life expectancy at birth, total (years)  SP.DYN.LE00.IN  32.535000   
3    Life expectancy at birth, total (years)  SP.DYN.LE00.IN  37.845152   
4    Life expectancy at birth, total (years)  SP.DYN.LE00.IN  38.211000   
..                                       ...             ...        ...   
261  Life expectancy at birth, total (years)  SP.DYN.LE00.IN  61.485000   
262  Life expectancy at birth, total (years)  SP.DYN.LE00.IN  33.678000   
263  Life expectancy at birth, total (years)  SP.DYN.LE00.IN  52.669000   
264  Life expectancy at birth, total (years)  SP.DYN.LE00.IN  49.042000   
265  Life expectancy at birth, total (years)  SP.DYN.LE00.IN  53.235000   

          1961       1962       1963       1964       1965  ...       2015  \
0    64.537000  64.752000  65.132000  65.294000  65.502000  ...  75.683000   
1    44.386697  44.752182  44.913159  45.479043  45.498338  ...  61.856458   
2    33.068000  33.547000  34.016000  34.494000  34.953000  ...  62.659000   
3    38.164950  38.735102  39.063715  39.335360  39.618038  ...  56.195872   
4    37.267000  37.539000  37.824000  38.131000  38.495000  ...  60.655000   
..         ...        ...        ...        ...        ...  ...        ...   
261  61.836000  62.134000  62.440000  62.734000  63.041000  ...  78.922000   
262  34.098000  33.615000  33.247000  34.738000  35.373000  ...  65.873000   
263  53.085000  53.376000  53.633000  53.906000  54.192000  ...  63.950000   
264  49.452000  49.794000  50.133000  49.849000  50.563000  ...  61.208000   
265  53.619000  54.071000  54.549000  54.994000  55.459000  ...  59.591000   

          2016       2017       2018       2019       2020       2021  \
0    75.617000  75.903000  76.072000  76.248000  75.723000  74.626000   
1    62.444050  62.922390  63.365863  63.755678  63.313860  62.454590   
2    63.136000  63.016000  63.081000  63.565000  62.575000  61.982000   
3    56.581678  56.888446  57.189139  57.555796  57.226373  56.988657   
4    61.092000  61.680000  62.144000  62.448000  62.261000  61.643000   
..         ...        ...        ...        ...        ...        ...   
261  78.981000  78.783000  78.696000  79.022000  76.567000  76.806000   
262  66.064000  65.957000  64.575000  65.092000  64.650000  63.753000   
263  64.747000  65.402000  65.674000  66.175000  65.252000  62.341000   
264  61.794000  62.120000  62.342000  62.793000  62.380000  61.223000   
265  60.306000  60.709000  61.414000  61.292000  61.124000  59.253000   

          2022  2023  Unnamed: 68  
0    74.992000   NaN          NaN  
1    62.899031   NaN          NaN  
2    62.879000   NaN          NaN  
3    57.626176   NaN          NaN  
4    61.929000   NaN          NaN  
..         ...   ...          ...  
261  79.524000   NaN          NaN  
262  63.720000   NaN          NaN  
263  61.480000   NaN          NaN  
264  61.803000   NaN          NaN  
265  59.391000   NaN          NaN  

[266 rows x 69 columns]

#### Q22. Drop the Country Code, Indicator Name, and Indicator Code columns. Then use .melt() to convert your data from wide to long. That is, instead of having one row per country and multiple colums per year, we want to have multiple rows per country and a single column for year. After melting, rename the columns to Country, Year, and Life_Expectancy.

In [153]:
life_expectancy = life_expectancy.drop(columns=['Country Code', 'Indicator Name', 'Indicator Code'])
life_expectancy

Country Name       1960       1961       1962       1963  \
0                          Aruba  64.152000  64.537000  64.752000  65.132000   
1    Africa Eastern and Southern  44.085552  44.386697  44.752182  44.913159   
2                    Afghanistan  32.535000  33.068000  33.547000  34.016000   
3     Africa Western and Central  37.845152  38.164950  38.735102  39.063715   
4                         Angola  38.211000  37.267000  37.539000  37.824000   
..                           ...        ...        ...        ...        ...   
261                       Kosovo  61.485000  61.836000  62.134000  62.440000   
262                  Yemen, Rep.  33.678000  34.098000  33.615000  33.247000   
263                 South Africa  52.669000  53.085000  53.376000  53.633000   
264                       Zambia  49.042000  49.452000  49.794000  50.133000   
265                     Zimbabwe  53.235000  53.619000  54.071000  54.549000   

          1964       1965       1966       1967       1968  ...       2015  \
0    65.294000  65.502000  66.063000  66.439000  66.757000  ...  75.683000   
1    45.479043  45.498338  45.249105  45.924905  46.223097  ...  61.856458   
2    34.494000  34.953000  35.453000  35.924000  36.418000  ...  62.659000   
3    39.335360  39.618038  39.837827  39.471500  40.085679  ...  56.195872   
4    38.131000  38.495000  38.757000  39.092000  39.484000  ...  60.655000   
..         ...        ...        ...        ...        ...  ...        ...   
261  62.734000  63.041000  63.323000  63.653000  63.935000  ...  78.922000   
262  34.738000  35.373000  36.097000  36.866000  37.796000  ...  65.873000   
263  53.906000  54.192000  54.391000  54.626000  54.876000  ...  63.950000   
264  49.849000  50.563000  50.679000  50.802000  50.856000  ...  61.208000   
265  54.994000  55.459000  55.889000  56.188000  56.455000  ...  59.591000   

          2016       2017       2018       2019       2020       2021  \
0    75.617000  75.903000  76.072000  76.248000  75.723000  74.626000   
1    62.444050  62.922390  63.365863  63.755678  63.313860  62.454590   
2    63.136000  63.016000  63.081000  63.565000  62.575000  61.982000   
3    56.581678  56.888446  57.189139  57.555796  57.226373  56.988657   
4    61.092000  61.680000  62.144000  62.448000  62.261000  61.643000   
..         ...        ...        ...        ...        ...        ...   
261  78.981000  78.783000  78.696000  79.022000  76.567000  76.806000   
262  66.064000  65.957000  64.575000  65.092000  64.650000  63.753000   
263  64.747000  65.402000  65.674000  66.175000  65.252000  62.341000   
264  61.794000  62.120000  62.342000  62.793000  62.380000  61.223000   
265  60.306000  60.709000  61.414000  61.292000  61.124000  59.253000   

          2022  2023  Unnamed: 68  
0    74.992000   NaN          NaN  
1    62.899031   NaN          NaN  
2    62.879000   NaN          NaN  
3    57.626176   NaN          NaN  
4    61.929000   NaN          NaN  
..         ...   ...          ...  
261  79.524000   NaN          NaN  
262  63.720000   NaN          NaN  
263  61.480000   NaN          NaN  
264  61.803000   NaN          NaN  
265  59.391000   NaN          NaN  

[266 rows x 66 columns]

In [155]:
life_expectancy = (
    life_expectancy
    .melt(id_vars = 'Country Name', var_name = 'Year', value_name = 'Life Expectancy')
    .rename(columns={'Country Name': 'Country'})
)
life_expectancy

Country         Year  Life Expectancy
0                            Aruba         1960        64.152000
1      Africa Eastern and Southern         1960        44.085552
2                      Afghanistan         1960        32.535000
3       Africa Western and Central         1960        37.845152
4                           Angola         1960        38.211000
...                            ...          ...              ...
17285                       Kosovo  Unnamed: 68              NaN
17286                  Yemen, Rep.  Unnamed: 68              NaN
17287                 South Africa  Unnamed: 68              NaN
17288                       Zambia  Unnamed: 68              NaN
17289                     Zimbabwe  Unnamed: 68              NaN

[17290 rows x 3 columns]

In [157]:
life_expectancy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17290 entries, 0 to 17289
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country          17290 non-null  object 
 1   Year             17290 non-null  object 
 2   Life Expectancy  16124 non-null  float64
dtypes: float64(1), object(2)
memory usage: 405.4+ KB


#### Q23. What was the first country with a life expectancy to exceed 80?

Japan, 1996

In [159]:
life_expectancy = life_expectancy.dropna()
life_expectancy

Country  Year  Life Expectancy
0                            Aruba  1960        64.152000
1      Africa Eastern and Southern  1960        44.085552
2                      Afghanistan  1960        32.535000
3       Africa Western and Central  1960        37.845152
4                           Angola  1960        38.211000
...                            ...   ...              ...
16753                       Kosovo  2022        79.524000
16754                  Yemen, Rep.  2022        63.720000
16755                 South Africa  2022        61.480000
16756                       Zambia  2022        61.803000
16757                     Zimbabwe  2022        59.391000

[16124 rows x 3 columns]

In [173]:
life_expectancy[life_expectancy['Life Expectancy'] >= 80]

Country  Year  Life Expectancy
9695                   Japan  1996        80.219756
9926               Gibraltar  1997        80.343000
9938    Hong Kong SAR, China  1997        80.112195
9961                   Japan  1997        80.424146
9988        Macao SAR, China  1997        80.162000
...                      ...   ...              ...
16692                  Qatar  2022        81.559000
16700              Singapore  2022        82.895122
16714               Slovenia  2022        81.282927
16715                 Sweden  2022        83.109756
16748  Virgin Islands (U.S.)  2022        80.319512

[690 rows x 3 columns]

#### Q24. Merge gdp_df and life_expectancy, keeping all countries and years that appear in both DataFrames. Save the result to a new DataFrame named gdp_le. If you get any errors in doing this, read them carefully and correct them. Look at the first five rows of your new data frame to confirm it merged correctly. Also, check the last five rows to make sure the data is clean and as expected.

In [181]:
life_expectancy['Year'] = life_expectancy['Year'].astype(int)

C:\Users\Maria\AppData\Local\Temp\ipykernel_12748\2397042444.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  life_expectancy['Year'] = life_expectancy['Year'].astype(int)


In [183]:
life_expectancy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16124 entries, 0 to 16757
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country          16124 non-null  object 
 1   Year             16124 non-null  int32  
 2   Life Expectancy  16124 non-null  float64
dtypes: float64(1), int32(1), object(1)
memory usage: 956.9+ KB


In [185]:
life_expectancy

Country  Year  Life Expectancy
0                            Aruba  1960        64.152000
1      Africa Eastern and Southern  1960        44.085552
2                      Afghanistan  1960        32.535000
3       Africa Western and Central  1960        37.845152
4                           Angola  1960        38.211000
...                            ...   ...              ...
16753                       Kosovo  2022        79.524000
16754                  Yemen, Rep.  2022        63.720000
16755                 South Africa  2022        61.480000
16756                       Zambia  2022        61.803000
16757                     Zimbabwe  2022        59.391000

[16124 rows x 3 columns]

In [197]:
gdp_le = pd.merge(gdp_df, life_expectancy, on = ['Country', 'Year'] , how = 'inner', validate='one_to_one')
gdp_le

Country  Year  GDP_Per_Capita Continent  Life Expectancy
0     Afghanistan  2021     1517.016266      Asia           61.982
1     Afghanistan  2020     1968.341002      Asia           62.575
2     Afghanistan  2019     2079.921861      Asia           63.565
3     Afghanistan  2018     2060.698973      Asia           63.081
4     Afghanistan  2017     2096.093111      Asia           63.016
...           ...   ...             ...       ...              ...
5499     Zimbabwe  1994     2670.106615    Africa           52.588
5500     Zimbabwe  1993     2458.783255    Africa           54.426
5501     Zimbabwe  1992     2468.278257    Africa           56.435
5502     Zimbabwe  1991     2781.787843    Africa           58.091
5503     Zimbabwe  1990     2704.757299    Africa           59.426

[5504 rows x 5 columns]

#### Q25. Create a new DataFrame, named gdp_le_2021 by extracting data for the year 2021 from gdp_le. How many countries have a life expectancy of at least 80 in 2021?